Misc net fixes

jukkar opened the pull request.

jukkar committed:
net: http: Yield after sending response in server

Make sure that a network packet is sent after calling
http_response_send_data(). Othwerwise the packets might be
piling up and not sent in timely manner.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print more TCP pending packet information

Print current list of pending packets if TCP debugging is
enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Cleanup TCP backlog when releasing context

If there is a backlog entry when TCP context is released, then
cancel the ACK timer if one exists.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Add CONFIG_NET_DEBUG_NET_PKT_ALL option

This option enables full debugging output for memory allocations.
As that can produce lot of output and slow down the device under test,
it is disabled by default.
The previous CONFIG_NET_DEBUG_NET_PKT will collect information about
memory allocations but will not print any output. Use "net mem" or
"net allocs" commands in net-shell to see the memory allocation status.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Set the TCP sent flag after really sending pkt

Make sure that the sent flag is only set after we have really
sent the packet and the driver has verified that.
If the net_pkt_set_sent() is called while still in tcp.c, then
depending on how fast the device is, it might happen that the
retry timer expires before the packet is actually sent. This was
seen in frdm-k64f with ethernet and various debug prints activated.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Add more debugging to TCP code

Useful when investigating issues.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Remove unused variable

There was two pkt variables in net_tcp_ack_received() so removing
the extra one.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -45,6 +45,7 @@
 #define NET_MAX_TCP_CONTEXT CONFIG_NET_MAX_CONTEXTS
 static struct net_tcp tcp_context[NET_MAX_TCP_CONTEXT];
 
+#define MIN_RETRY_MS 5000
```
Just realized that this is wrong fix. I will abandon this specific commit.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -108,6 +108,9 @@ static bool net_if_tx(struct net_if *iface)
 #if defined(CONFIG_NET_STATISTICS)
 		pkt_len = net_pkt_get_len(pkt);
 #endif
+#if defined(CONFIG_NET_TCP)
```
This commit needs more TLC. The various do_ref_if_needed() calls in tcp.c must to be done **only** if we could actually sent the packet.

tbursztyka replied:
what about using IF_ENABLED() macro? It's cleaner than #ifdef/#endif (if usable).

jukkar replied:
Sure, I send a new version

jukkar reviewed (commented):
I will send new version of this one.

jukkar committed:
net: tcp: Only do ref when resending if we could send the packet

If the packet sending is slow then we must NOT increment the ref
count when re-sending it. This is unlikely but can happen if there
are lot of debug prints etc. extra activities that prevent the driver
to actually send the packet fast enough.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
The commit "net: http: Yield after sending response in server" looks varying. Is there formal explanation why it's needed? Aren't we on preemptive multitasking? Maybe we don't have the priorities right?

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -626,6 +626,23 @@ int net_context_unref(struct net_context *context)
+		int i;
+
+		/* Clear the backlog for this TCP context. */
+		for (i = 0; i < CONFIG_NET_TCP_BACKLOG_SIZE; i++) {
```
I asked this before, let me ask again: shouldn't we protect access to tcp_backlog[]? Can't the delayed work preempt our code?

jukkar commented:
> The commit "net: http: Yield after sending response in server" looks varying. Is there formal explanation why it's needed? Aren't we on preemptive multitasking? Maybe we don't have the priorities right?

If someone starts to send network packets and is running in co-op thread, then it might happen that network TX thread does not have chance to run and the sending packets just pile up in TX queue. So just in case we can call k_yield() here after sending the HTTP packet so that we make sure that the packet gets sent asap.

jukkar committed:
net: shell: Fix column alignment in net mem command

Because the name of the memory pool can be quite long, print
it last so that the columns get aligned nicely.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Do not re-send already sent packets

If the expire send timer expires, then it sends the packet.
If that happens, then we must not try to send the same packet
again if we receive ACK etc. which can cause re-sends to happen.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tcp: Send data only in established state

If we are not in ESTABLISHED state, then there is no need to
try to resend any pending data packets.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented on subsys/net/ip/Kconfig.debug:
```diff
@@ -91,7 +91,16 @@ config NET_DEBUG_NET_PKT
+	Enables collection of network packet and buffer allocations and frees.
+
+config NET_DEBUG_NET_PKT_ALL
+	bool "Debug network packet and buffer individual allocation"
```
I'm no sure that from title alone it's clear what this option does vs NET_DEBUG_NET_PKT. Perhaps it can be clarified.

tbursztyka replied:
That reminds me I actually ask for better naming (though indeed makes the patch more complicated). NET_DEBUG_NET_PKT could be kept, and would be the one printing out everything. But an additional NET_DEBUG_NET_PKT_MEM_STATS (for instance) could be introduced and would act as collecting net pkt/buf allocs and frees.

jukkar replied:
My argument for the proposed patch is:

* existing NET_DEBUG_NET_PKT is kept but it is only collecting memory allocation information
* new NET_DEBUG_NET_PKT_ALL will print all the allocations, it depends on NET_DEBUG_NET_PKT

With these changes, we do not need to change much code. I tried the proposal that @tbursztyka had, but to my eyes this required more changes.

pfalcon commented:
Some commits look great (love more debug output), some raise comments above. It's a bit hard to review so wide varying commits in one PR.

jukkar commented:
> Some commits look great (love more debug output), some raise comments above. It's a bit hard to review so wide varying commits in one PR.

With these changes I managed to get TCP finally working, as you can probably see from many TCP changes introduced here.

@pfalcon The issue with backlog locking might go away with #1141

jukkar commented:
While rebasing the set, I seem to have managed to cause some issues. I will send a fixed version.

jukkar commented:
> It's a bit hard to review so wide varying commits in one PR.

These were the issues I saw when working with #980 
Fortunately the commits are rather small here, I can send these in separate PRs if really needed but tried to keep this set like this in order to avoid too much work in my side.

Vudentz commented on subsys/net/ip/net_context.c:
```diff
@@ -258,7 +258,7 @@ static int tcp_backlog_ack(struct net_pkt *pkt, struct net_context *context)
 	context->tcp->send_ack = tcp_backlog[r].send_ack;
 
-	if (k_delayed_work_cancel(&tcp_backlog[r].ack_timer) < 0) {
+	if (k_delayed_work_cancel(&tcp_backlog[r].ack_timer) == -EINPROGRESS) {
```
Perhaps we should got with https://github.com/zephyrproject-rtos/zephyr/pull/1141 since k_delayed_work_cancel no longer return -EINPROGRESS, and -EAGAIN is fine since it just mean the work is idle.

jukkar replied:
Sure, I can drop that patch from this set.

pfalcon reviewed (approved):
> I can send these in separate PRs if really needed

Nope, it's just my usual "complaint" ;-).
